File: advanced/src/main/kotlin/com/diego/kotlincourse/classes/NestedClasses.java
package com.diego.kotlincourse.classes;

public class NestedClasses {
    public static void main(String[] args) {
        //DirectoryExplorer.PermissionCheck pc = new DirectoryExplorer.PermissionCheck();
        //Log log = Log.Factory.createFileLog("file.txt");
        Log log = Log.createFileLog("file.txt");
        //Log logDirectly = Log() // we can't access either
    }
}

File: advanced/src/main/kotlin/com/diego/kotlincourse/classes/NestedClasses.kt
package com.diego.kotlincourse.classes

class DirectoryExplorer(val user: String = "") {
    fun listFolder(folder: String, user: String){
        val permissionCheck = PermissionCheck()
        permissionCheck.validatePermission()
    }

    inner class PermissionCheck {
        fun validatePermission() {
            if(user != "diego"){

            }
        }

    }
}

fun main(args: Array<String>) {
    val de = DirectoryExplorer("diego")
    //val pc = DirectoryExplorer.PermissionCheck()
    val pc = DirectoryExplorer().PermissionCheck() // when using 'inner'
}